File: java/1306_JumpGameIII.java
class Solution {
    public boolean canReach(int[] arr, int start) {
        int n = arr.length;
        Queue<Integer> q = new LinkedList<Integer>() {{
            add(start);
        }};
        
        while (!q.isEmpty()) {
            int idx = q.poll();
            if (arr[idx] == 0) return true;
            else if (arr[idx] > 0) {
                if (idx - arr[idx] >= 0) q.add(idx - arr[idx]);
                if (idx + arr[idx] < n) q.add(idx + arr[idx]);
                arr[idx] = -arr[idx];
            }
        }
        return false;
    }
}

File: cpp/1306_JumpGameIII.cpp
class Solution {
public:
    bool canReach(vector<int>& arr, int start) {
        return dfs(arr, start);
    }

    bool dfs(vector<int>& arr, int idx) {
        if (idx < 0 || idx >= arr.size() || arr[idx] < 0) {
            return false;
        } else if (arr[idx] == 0) {
            return true;
        }

        arr[idx] = -arr[idx];
        return dfs(arr, idx - arr[idx]) || dfs(arr, idx + arr[idx]);
    }
};
